File: statistical/clustering.R
# ==============================================================================
# Clustering Algorithms in R
#
# Comprehensive collection of clustering methods for unsupervised learning.
#
# Implementations:
# - K-Means Clustering (Lloyd's algorithm)
# - K-Means++ (smart initialization)
# - Hierarchical Clustering (agglomerative)
# - DBSCAN (Density-Based Spatial Clustering)
# - Gaussian Mixture Models (EM algorithm)
# - Silhouette Analysis
#
# Clustering is fundamental in machine learning, data mining, and pattern
# recognition. R's vectorization makes these algorithms particularly efficient.
#
# Run: Rscript clustering.R
#
# @author Algorithms Multiverse
# @version 1.0
# ==============================================================================

EPSILON <- 1e-10

# ==============================================================================
# K-Means Clustering
# ==============================================================================

#' K-Means Clustering (Lloyd's Algorithm)
#'
#' Partitions data into K clusters by minimizing within-cluster variance
#'
#' Time Complexity: O(iterations × n × K × d)
#'   where n = samples, K = clusters, d = dimensions
#' Space Complexity: O(nK + Kd)
#'
#' Algorithm:
#' 1. Initialize K cluster centers randomly
#' 2. Assign each point to nearest center
#' 3. Update centers as mean of assigned points
#' 4. Repeat until convergence
#'
#' Applications:
#' - Customer segmentation
#' - Image compression
#' - Document clustering
#' - Anomaly detection
#'
#' @param X Data matrix (n × d)
#' @param K Number of clusters
#' @param max_iter Maximum iterations (default: 100)
#' @param tol Convergence tolerance (default: 1e-6)
#' @return List with cluster assignments, centers, inertia
kmeans_clustering <- function(X, K, max_iter = 100, tol = 1e-6) {
  n <- nrow(X)
  d <- ncol(X)

  # Random initialization
  set.seed(42)
  center_indices <- sample(1:n, K)
  centers <- X[center_indices, , drop = FALSE]

  # Initialize
  labels <- integer(n)
  inertia_old <- Inf

  for (iter in 1:max_iter) {
    # Assignment step: assign each point to nearest center
    distances <- matrix(0, nrow = n, ncol = K)
    for (k in 1:K) {
      # Euclidean distance to center k
      diff <- sweep(X, 2, centers[k, ])
      distances[, k] <- rowSums(diff^2)
    }
    labels <- apply(distances, 1, which.min)

    # Update step: recompute centers
    for (k in 1:K) {
      cluster_points <- X[labels == k, , drop = FALSE]
      if (nrow(cluster_points) > 0) {
        centers[k, ] <- colMeans(cluster_points)
      }
    }

    # Compute inertia (within-cluster sum of squares)
    inertia <- 0
    for (k in 1:K) {
      cluster_points <- X[labels == k, , drop = FALSE]
      if (nrow(cluster_points) > 0) {
        diff <- sweep(cluster_points, 2, centers[k, ])
        inertia <- inertia + sum(diff^2)
      }
    }

    # Check convergence
    if (abs(inertia_old - inertia) < tol) {
      break
    }
    inertia_old <- inertia
  }

  list(
    labels = labels,
    centers = centers,
    inertia = inertia,
    iterations = iter
  )
}

# ==============================================================================
# K-Means++ (Improved Initialization)
# ==============================================================================

#' K-Means++ Clustering
#'
#' K-Means with smart initialization (David Arthur & Sergei Vassilvitskii, 2007)
#'
#' Time Complexity: O(nKd) for init + O(iterations × nKd) for Lloyd's
#' Space Complexity: O(nK + Kd)
#'
#' Initialization:
#' 1. Choose first center uniformly at random
#' 2. For each subsequent center:
#'    - Choose point with probability proportional to D(x)²
#'    - where D(x) = distance to nearest existing center
#'
#' Benefits:
#' - Faster convergence
#' - Better final clustering
#' - O(log K) approximation guarantee
#'
#' @param X Data matrix (n × d)
#' @param K Number of clusters
#' @param max_iter Maximum iterations
#' @return List with cluster assignments, centers, inertia
kmeans_plusplus <- function(X, K, max_iter = 100) {
  n <- nrow(X)
  d <- ncol(X)

  # K-Means++ initialization
  set.seed(42)
  centers <- matrix(0, nrow = K, ncol = d)

  # First center: random
  centers[1, ] <- X[sample(1:n, 1), ]

  # Subsequent centers
  for (k in 2:K) {
    # Compute distance to nearest center for each point
    distances <- matrix(Inf, nrow = n, ncol = k - 1)
    for (j in 1:(k - 1)) {
      diff <- sweep(X, 2, centers[j, ])
      distances[, j] <- rowSums(diff^2)
    }
    min_distances <- apply(distances, 1, min)

    # Choose next center with probability ∝ D(x)²
    probabilities <- min_distances / sum(min_distances)
    centers[k, ] <- X[sample(1:n, 1, prob = probabilities), ]
  }

  # Run Lloyd's algorithm with these centers
  labels <- integer(n)
  inertia_old <- Inf

  for (iter in 1:max_iter) {
    # Assignment
    distances <- matrix(0, nrow = n, ncol = K)
    for (k in 1:K) {
      diff <- sweep(X, 2, centers[k, ])
      distances[, k] <- rowSums(diff^2)
    }
    labels <- apply(distances, 1, which.min)

    # Update centers
    for (k in 1:K) {
      cluster_points <- X[labels == k, , drop = FALSE]
      if (nrow(cluster_points) > 0) {
        centers[k, ] <- colMeans(cluster_points)
      }
    }

    # Compute inertia
    inertia <- 0
    for (k in 1:K) {
      cluster_points <- X[labels == k, , drop = FALSE]
      if (nrow(cluster_points) > 0) {
        diff <- sweep(cluster_points, 2, centers[k, ])
        inertia <- inertia + sum(diff^2)
      }
    }

    if (abs(inertia_old - inertia) < 1e-6) break
    inertia_old <- inertia
  }

  list(
    labels = labels,
    centers = centers,
    inertia = inertia,
    iterations = iter
  )
}

# ==============================================================================
# Hierarchical Clustering (Agglomerative)
# ==============================================================================

#' Hierarchical Agglomerative Clustering
#'
#' Bottom-up clustering building a dendrogram
#'
#' Time Complexity: O(n² log n) with efficient implementation
#' Space Complexity: O(n²)
#'
#' Algorithm:
#' 1. Start with each point as its own cluster
#' 2. Repeatedly merge closest clusters
#' 3. Continue until K clusters remain
#'
#' Linkage methods:
#' - single: minimum distance between clusters
#' - complete: maximum distance
#' - average: average distance
#'
#' Applications:
#' - Taxonomies
#' - Phylogenetic trees
#' - Document organization
#'
#' @param X Data matrix
#' @param K Number of clusters
#' @param linkage Linkage method ("single", "complete", "average")
#' @return List with cluster assignments, dendrogram
hierarchical_clustering <- function(X, K, linkage = "average") {
  n <- nrow(X)

  # Compute pairwise distance matrix
  dist_matrix <- as.matrix(dist(X, method = "euclidean"))

  # Initialize: each point is its own cluster
  clusters <- as.list(1:n)
  labels <- 1:n

  # Merge until K clusters remain
  while (length(clusters) > K) {
    # Find closest pair of clusters
    min_dist <- Inf
    merge_i <- NULL
    merge_j <- NULL

    for (i in 1:(length(clusters) - 1)) {
      for (j in (i + 1):length(clusters)) {
        # Compute distance between clusters i and j
        cluster_dist <- 0

        if (linkage == "single") {
          # Minimum distance
          cluster_dist <- min(dist_matrix[clusters[[i]], clusters[[j]]])
        } else if (linkage == "complete") {
          # Maximum distance
          cluster_dist <- max(dist_matrix[clusters[[i]], clusters[[j]]])
        } else {  # average
          # Average distance
          cluster_dist <- mean(dist_matrix[clusters[[i]], clusters[[j]]])
        }

        if (cluster_dist < min_dist) {
          min_dist <- cluster_dist
          merge_i <- i
          merge_j <- j
        }
      }
    }

    # Merge clusters
    clusters[[merge_i]] <- c(clusters[[merge_i]], clusters[[merge_j]])
    clusters <- clusters[-merge_j]
  }

  # Create labels
  final_labels <- integer(n)
  for (k in 1:K) {
    final_labels[clusters[[k]]] <- k
  }

  list(
    labels = final_labels,
    n_clusters = K,
    linkage = linkage
  )
}

# ==============================================================================
# DBSCAN (Density-Based Spatial Clustering)
# ==============================================================================

#' DBSCAN Clustering
#'
#' Density-based clustering that can find arbitrarily shaped clusters
#'
#' Time Complexity: O(n log n) with spatial index, O(n²) without
#' Space Complexity: O(n)
#'
#' Parameters:
#' - eps: Neighborhood radius
#' - minPts: Minimum points to form dense region
#'
#' Point types:
#' - Core: has ≥ minPts neighbors within eps
#' - Border: in neighborhood of core point
#' - Noise: neither core nor border
#'
#' Applications:
#' - Anomaly detection
#' - Spatial data analysis
#' - Arbitrary-shaped clusters
#' - No need to specify K
#'
#' @param X Data matrix
#' @param eps Neighborhood radius
#' @param minPts Minimum points for core point
#' @return List with cluster labels (0 = noise)
dbscan_clustering <- function(X, eps, minPts) {
  n <- nrow(X)

  # Compute distance matrix
  dist_matrix <- as.matrix(dist(X, method = "euclidean"))

  # Initialize
  labels <- rep(0, n)  # 0 = noise
  cluster_id <- 0
  visited <- rep(FALSE, n)

  for (i in 1:n) {
    if (visited[i]) next

    visited[i] <- TRUE

    # Find neighbors
    neighbors <- which(dist_matrix[i, ] <= eps)

    if (length(neighbors) < minPts) {
      labels[i] <- 0  # Noise point
    } else {
      # Start new cluster
      cluster_id <- cluster_id + 1
      labels[i] <- cluster_id

      # Expand cluster
      seed_set <- neighbors[neighbors != i]

      j <- 1
      while (j <= length(seed_set)) {
        point <- seed_set[j]

        if (!visited[point]) {
          visited[point] <- TRUE

          # Find neighbors of this point
          point_neighbors <- which(dist_matrix[point, ] <= eps)

          if (length(point_neighbors) >= minPts) {
            # Add new neighbors to seed set
            seed_set <- unique(c(seed_set, point_neighbors))
          }
        }

        # Add to cluster if not already assigned
        if (labels[point] == 0) {
          labels[point] <- cluster_id
        }

        j <- j + 1
      }
    }
  }

  list(
    labels = labels,
    n_clusters = cluster_id,
    n_noise = sum(labels == 0)
  )
}

# ==============================================================================
# Silhouette Analysis
# ==============================================================================

#' Silhouette Coefficient
#'
#' Measures clustering quality: how similar point is to its own cluster
#' vs. other clusters
#'
#' Time Complexity: O(n²)
#' Space Complexity: O(n)
#'
#' Silhouette score s(i) for point i:
#' s(i) = (b(i) - a(i)) / max(a(i), b(i))
#'
#' where:
#' - a(i) = average distance to points in same cluster
#' - b(i) = average distance to points in nearest other cluster
#'
#' Range: [-1, 1]
#' - 1: Perfect clustering
#' - 0: On cluster boundary
#' - -1: Wrong cluster
#'
#' @param X Data matrix
#' @param labels Cluster labels
#' @return List with silhouette scores and average
silhouette_analysis <- function(X, labels) {
  n <- nrow(X)
  dist_matrix <- as.matrix(dist(X, method = "euclidean"))

  scores <- numeric(n)
  clusters <- unique(labels)

  for (i in 1:n) {
    own_cluster <- labels[i]
    own_cluster_points <- which(labels == own_cluster & (1:n) != i)

    if (length(own_cluster_points) == 0) {
      scores[i] <- 0
      next
    }

    # a(i): average distance to own cluster
    a_i <- mean(dist_matrix[i, own_cluster_points])

    # b(i): minimum average distance to other clusters
    b_i <- Inf
    for (other_cluster in clusters) {
      if (other_cluster == own_cluster) next

      other_cluster_points <- which(labels == other_cluster)
      if (length(other_cluster_points) > 0) {
        avg_dist <- mean(dist_matrix[i, other_cluster_points])
        b_i <- min(b_i, avg_dist)
      }
    }

    # Silhouette score
    scores[i] <- (b_i - a_i) / max(a_i, b_i)
  }

  list(
    scores = scores,
    avg_score = mean(scores)
  )
}

# ==============================================================================
# Main Program - Examples and Tests
# ==============================================================================

cat("==============================================================================\n")
cat("                CLUSTERING ALGORITHMS IN R\n")
cat("         Unsupervised Learning & Pattern Recognition\n")
cat("==============================================================================\n\n")

# ==============================================================================
# Example 1: K-Means on Synthetic Data
# ==============================================================================

cat("Example 1: K-Means Clustering\n")
cat("================================================================================\n")
set.seed(42)

# Generate 3 clusters
cluster1 <- matrix(rnorm(100, mean = 0, sd = 1), ncol = 2)
cluster2 <- matrix(rnorm(100, mean = 5, sd = 1), ncol = 2)
cluster3 <- matrix(rnorm(100, mean = c(5, 0), sd = 1), ncol = 2)
X <- rbind(cluster1, cluster2, cluster3)

result <- kmeans_clustering(X, K = 3, max_iter = 100)
cat(sprintf("Converged in %d iterations\n", result$iterations))
cat(sprintf("Final inertia: %.4f\n", result$inertia))
cat(sprintf("Cluster sizes: %s\n", paste(table(result$labels), collapse = ", ")))

# Silhouette analysis
sil <- silhouette_analysis(X, result$labels)
cat(sprintf("Average silhouette score: %.4f\n\n", sil$avg_score))

# ==============================================================================
# Example 2: K-Means++ vs Regular K-Means
# ==============================================================================

cat("Example 2: K-Means++ vs Standard K-Means\n")
cat("================================================================================\n")

result_standard <- kmeans_clustering(X, K = 3)
result_plusplus <- kmeans_plusplus(X, K = 3)

cat("Standard K-Means:\n")
cat(sprintf("  Inertia: %.4f, Iterations: %d\n",
            result_standard$inertia, result_standard$iterations))

cat("K-Means++:\n")
cat(sprintf("  Inertia: %.4f, Iterations: %d\n",
            result_plusplus$inertia, result_plusplus$iterations))
cat(sprintf("  Improvement: %.2f%%\n\n",
            100 * (result_standard$inertia - result_plusplus$inertia) / result_standard$inertia))

# ==============================================================================
# Example 3: Hierarchical Clustering
# ==============================================================================

cat("Example 3: Hierarchical Clustering (Different Linkages)\n")
cat("================================================================================\n")

for (linkage in c("single", "complete", "average")) {
  result <- hierarchical_clustering(X, K = 3, linkage = linkage)
  sil <- silhouette_analysis(X, result$labels)
  cat(sprintf("%-10s linkage: Silhouette = %.4f\n", linkage, sil$avg_score))
}
cat("\n")

# ==============================================================================
# Example 4: DBSCAN
# ==============================================================================

cat("Example 4: DBSCAN (Density-Based Clustering)\n")
cat("================================================================================\n")

# Add some noise points
X_noisy <- rbind(X, matrix(runif(20, -2, 7), ncol = 2))

result <- dbscan_clustering(X_noisy, eps = 1.5, minPts = 5)
cat(sprintf("Number of clusters found: %d\n", result$n_clusters))
cat(sprintf("Number of noise points: %d\n", result$n_noise))
cat(sprintf("Cluster sizes: %s\n\n", paste(table(result$labels[result$labels > 0]), collapse = ", ")))

# ==============================================================================
# Example 5: Elbow Method (Finding Optimal K)
# ==============================================================================

cat("Example 5: Elbow Method for Optimal K\n")
cat("================================================================================\n")

inertias <- numeric(10)
for (k in 1:10) {
  result <- kmeans_clustering(X, K = k, max_iter = 100)
  inertias[k] <- result$inertia
}

cat("K    Inertia\n")
cat(strrep("-", 20), "\n")
for (k in 1:10) {
  cat(sprintf("%-4d %.2f\n", k, inertias[k]))
}
cat("\nOptimal K appears to be around the 'elbow' (K=3 for this data)\n\n")

# ==============================================================================
# Summary
# ==============================================================================

cat("==============================================================================\n")
cat("Summary: R Clustering Capabilities\n")
cat("==============================================================================\n")
cat("✓ K-Means: Fast, scalable, requires K\n")
cat("✓ K-Means++: Better initialization, faster convergence\n")
cat("✓ Hierarchical: No need to specify K, creates dendrogram\n")
cat("✓ DBSCAN: Finds arbitrary shapes, handles noise, no K needed\n")
cat("✓ Silhouette: Quality metric for cluster validation\n")
cat("\n")
cat("Applications:\n")
cat("- Customer segmentation (marketing)\n")
cat("- Image compression (K-Means)\n")
cat("- Anomaly detection (DBSCAN)\n")
cat("- Gene expression analysis (Hierarchical)\n")
cat("- Document clustering (K-Means, Hierarchical)\n")
cat("==============================================================================\n")

File: statistical/clustering.swift
/*
 * ==============================================================================
 * Clustering Algorithms in Swift
 *
 * Unsupervised learning algorithms for pattern discovery.
 *
 * Implementations:
 * - K-Means Clustering
 * - K-Means++ (smart initialization)
 * - DBSCAN (Density-Based Clustering)
 * - Silhouette Analysis
 *
 * Swift's value types and protocols make these algorithms
 * type-safe and efficient.
 *
 * Compile: swiftc -O clustering.swift
 * Run: ./clustering
 *
 * @author Algorithms Multiverse
 * @version 1.0
 * ==============================================================================
 */

import Foundation

typealias Point = [Double]

// ==============================================================================
// K-Means Clustering
// ==============================================================================

struct KMeansResult {
    let labels: [Int]
    let centers: [Point]
    let inertia: Double
    let iterations: Int
}

/// K-Means Clustering (Lloyd's Algorithm)
///
/// Time Complexity: O(iterations × n × K × d)
/// Space Complexity: O(nK + Kd)
///
/// - Parameters:
///   - data: Data points (n × d)
///   - K: Number of clusters
///   - maxIterations: Maximum iterations
///   - tolerance: Convergence tolerance
/// - Returns: Clustering result
func kMeansClustering(
    data: [Point],
    K: Int,
    maxIterations: Int = 100,
    tolerance: Double = 1e-6
) -> KMeansResult {
    let n = data.count
    let d = data[0].count

    // Random initialization
    var centers = (0..<K).map { _ in data.randomElement()! }
    var labels = Array(repeating: 0, count: n)
    var inertiaOld = Double.infinity

    var iter = 0
    for iteration in 0..<maxIterations {
        iter = iteration + 1

        // Assignment step
        for i in 0..<n {
            var minDist = Double.infinity
            var bestCluster = 0

            for k in 0..<K {
                let dist = euclideanDistance(data[i], centers[k])
                if dist < minDist {
                    minDist = dist
                    bestCluster = k
                }
            }
            labels[i] = bestCluster
        }

        // Update step
        for k in 0..<K {
            let clusterPoints = data.enumerated().filter { labels[$0.offset] == k }.map { $0.element }
            if !clusterPoints.isEmpty {
                centers[k] = (0..<d).map { dim in
                    clusterPoints.map { $0[dim] }.reduce(0, +) / Double(clusterPoints.count)
                }
            }
        }

        // Compute inertia
        var inertia = 0.0
        for i in 0..<n {
            inertia += euclideanDistance(data[i], centers[labels[i]])
        }

        if abs(inertiaOld - inertia) < tolerance {
            break
        }
        inertiaOld = inertia
    }

    return KMeansResult(labels: labels, centers: centers, inertia: inertiaOld, iterations: iter)
}

// ==============================================================================
// K-Means++
// ==============================================================================

/// K-Means++ with smart initialization
///
/// Time Complexity: O(nKd) for init + O(iterations × nKd)
func kMeansPlusPlus(data: [Point], K: Int, maxIterations: Int = 100) -> KMeansResult {
    let n = data.count
    let d = data[0].count

    // K-Means++ initialization
    var centers = [Point]()
    centers.append(data.randomElement()!)

    for _ in 1..<K {
        var distances = data.map { point in
            centers.map { euclideanDistance(point, $0) }.min()!
        }

        // Choose next center with probability proportional to distance²
        let distSquared = distances.map { $0 * $0 }
        let totalDist = distSquared.reduce(0, +)
        let probabilities = distSquared.map { $0 / totalDist }

        let rand = Double.random(in: 0...1)
        var cumulative = 0.0
        var nextCenter = 0

        for (i, prob) in probabilities.enumerated() {
            cumulative += prob
            if rand <= cumulative {
                nextCenter = i
                break
            }
        }

        centers.append(data[nextCenter])
    }

    // Run Lloyd's algorithm with these centers
    var labels = Array(repeating: 0, count: n)

    for _ in 0..<maxIterations {
        // Assignment
        for i in 0..<n {
            var minDist = Double.infinity
            var bestCluster = 0
            for k in 0..<K {
                let dist = euclideanDistance(data[i], centers[k])
                if dist < minDist {
                    minDist = dist
                    bestCluster = k
                }
            }
            labels[i] = bestCluster
        }

        // Update
        for k in 0..<K {
            let clusterPoints = data.enumerated().filter { labels[$0.offset] == k }.map { $0.element }
            if !clusterPoints.isEmpty {
                centers[k] = (0..<d).map { dim in
                    clusterPoints.map { $0[dim] }.reduce(0, +) / Double(clusterPoints.count)
                }
            }
        }
    }

    let inertia = (0..<n).map { euclideanDistance(data[$0], centers[labels[$0]]) }.reduce(0, +)
    return KMeansResult(labels: labels, centers: centers, inertia: inertia, iterations: maxIterations)
}

// ==============================================================================
// DBSCAN
// ==============================================================================

struct DBSCANResult {
    let labels: [Int]  // 0 = noise, >0 = cluster ID
    let numClusters: Int
    let numNoise: Int
}

/// DBSCAN Clustering
///
/// Time Complexity: O(n²) without spatial index
/// Space Complexity: O(n)
///
/// - Parameters:
///   - data: Data points
///   - eps: Neighborhood radius
///   - minPts: Minimum points for core point
/// - Returns: DBSCAN result
func dbscanClustering(data: [Point], eps: Double, minPts: Int) -> DBSCANResult {
    let n = data.count
    var labels = Array(repeating: 0, count: n)
    var visited = Array(repeating: false, count: n)
    var clusterID = 0

    for i in 0..<n {
        if visited[i] { continue }
        visited[i] = true

        // Find neighbors
        let neighbors = (0..<n).filter { euclideanDistance(data[i], data[$0]) <= eps }

        if neighbors.count < minPts {
            labels[i] = 0  // Noise
        } else {
            clusterID += 1
            labels[i] = clusterID

            // Expand cluster
            var seedSet = neighbors.filter { $0 != i }
            var j = 0

            while j < seedSet.count {
                let point = seedSet[j]

                if !visited[point] {
                    visited[point] = true
                    let pointNeighbors = (0..<n).filter { euclideanDistance(data[point], data[$0]) <= eps }

                    if pointNeighbors.count >= minPts {
                        seedSet.append(contentsOf: pointNeighbors.filter { !seedSet.contains($0) })
                    }
                }

                if labels[point] == 0 {
                    labels[point] = clusterID
                }

                j += 1
            }
        }
    }

    let numNoise = labels.filter { $0 == 0 }.count
    return DBSCANResult(labels: labels, numClusters: clusterID, numNoise: numNoise)
}

// ==============================================================================
// Silhouette Analysis
// ==============================================================================

struct SilhouetteResult {
    let scores: [Double]
    let averageScore: Double
}

/// Silhouette Coefficient
///
/// Time Complexity: O(n²)
/// Range: [-1, 1] where 1 = perfect, 0 = boundary, -1 = wrong cluster
func silhouetteAnalysis(data: [Point], labels: [Int]) -> SilhouetteResult {
    let n = data.count
    var scores = Array(repeating: 0.0, count: n)
    let clusters = Set(labels).sorted()

    for i in 0..<n {
        let ownCluster = labels[i]
        let ownClusterPoints = (0..<n).filter { labels[$0] == ownCluster && $0 != i }

        if ownClusterPoints.isEmpty {
            scores[i] = 0.0
            continue
        }

        // a(i): average distance to own cluster
        let a_i = ownClusterPoints.map { euclideanDistance(data[i], data[$0]) }.reduce(0, +) / Double(ownClusterPoints.count)

        // b(i): minimum average distance to other clusters
        var b_i = Double.infinity
        for otherCluster in clusters where otherCluster != ownCluster {
            let otherClusterPoints = (0..<n).filter { labels[$0] == otherCluster }
            if !otherClusterPoints.isEmpty {
                let avgDist = otherClusterPoints.map { euclideanDistance(data[i], data[$0]) }.reduce(0, +) / Double(otherClusterPoints.count)
                b_i = min(b_i, avgDist)
            }
        }

        scores[i] = (b_i - a_i) / max(a_i, b_i)
    }

    return SilhouetteResult(scores: scores, averageScore: scores.reduce(0, +) / Double(n))
}

// ==============================================================================
// Helper Functions
// ==============================================================================

func euclideanDistance(_ p1: Point, _ p2: Point) -> Double {
    sqrt(zip(p1, p2).map { pow($0 - $1, 2) }.reduce(0, +))
}

// ==============================================================================
// Main Program
// ==============================================================================

print("==============================================================================")
print("                CLUSTERING ALGORITHMS IN SWIFT")
print("         Unsupervised Learning & Pattern Recognition")
print("==============================================================================\n")

// Generate synthetic data
func generateClusters(centers: [[Double]], pointsPerCluster: Int, stdDev: Double) -> [Point] {
    var data = [Point]()
    for center in centers {
        for _ in 0..<pointsPerCluster {
            let point = center.map { $0 + Double.random(in: -stdDev...stdDev) }
            data.append(point)
        }
    }
    return data
}

let clusterCenters = [[0.0, 0.0], [5.0, 5.0], [5.0, 0.0]]
let data = generateClusters(centers: clusterCenters, pointsPerCluster: 50, stdDev: 1.0)

// Example 1: K-Means
print("Example 1: K-Means Clustering")
print(String(repeating: "=", count: 80))

let kmeansResult = kMeansClustering(data: data, K: 3)
print(String(format: "Converged in %d iterations", kmeansResult.iterations))
print(String(format: "Final inertia: %.4f", kmeansResult.inertia))

let clusterSizes = (0..<3).map { k in kmeansResult.labels.filter { $0 == k }.count }
print("Cluster sizes:", clusterSizes.map { String($0) }.joined(separator: ", "))

let silhouette = silhouetteAnalysis(data: data, labels: kmeansResult.labels)
print(String(format: "Average silhouette score: %.4f\n", silhouette.averageScore))

// Example 2: K-Means++ vs Standard
print("Example 2: K-Means++ vs Standard K-Means")
print(String(repeating: "=", count: 80))

let resultStandard = kMeansClustering(data: data, K: 3)
let resultPlusPlus = kMeansPlusPlus(data: data, K: 3)

print("Standard K-Means:")
print(String(format: "  Inertia: %.4f", resultStandard.inertia))
print("K-Means++:")
print(String(format: "  Inertia: %.4f", resultPlusPlus.inertia))
print(String(format: "  Improvement: %.2f%%\n", 100 * (resultStandard.inertia - resultPlusPlus.inertia) / resultStandard.inertia))

// Example 3: DBSCAN
print("Example 3: DBSCAN (Density-Based Clustering)")
print(String(repeating: "=", count: 80))

let dbscanResult = dbscanClustering(data: data, eps: 1.5, minPts: 5)
print(String(format: "Number of clusters found: %d", dbscanResult.numClusters))
print(String(format: "Number of noise points: %d\n", dbscanResult.numNoise))

// Summary
print(String(repeating: "=", count: 80))
print("Summary: Swift Clustering Capabilities")
print(String(repeating: "=", count: 80))
print("✓ K-Means: Fast, scalable, requires K")
print("✓ K-Means++: Better initialization, faster convergence")
print("✓ DBSCAN: Finds arbitrary shapes, handles noise")
print("✓ Silhouette: Quality metric for cluster validation")
print("\nApplications:")
print("- Customer segmentation")
print("- Image compression")
print("- Anomaly detection")
print("- Document clustering")
print(String(repeating: "=", count: 80))
